dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,169 @@
+    auto &decoder_data = resource.decoder_data;
+    auto backend = NVJPEG_BACKEND_HYBRID;  // TODO(msala) allow other backens
+    CUDA_CALL(nvjpegDecoderCreate(nvjpeg_handle_, backend, &decoder_data.decoder));
+    CUDA_CALL(nvjpegDecoderStateCreate(nvjpeg_handle_, decoder_data.decoder, &decoder_data.state));
+  }
+}
+
+NvJpegDecoderInstance::~NvJpegDecoderInstance() {
+  for (auto &thread_id : tp_->GetThreadIds()) {
+    nvjpeg_memory::DeleteAllBuffers(thread_id);
+  }
+
+  for (const auto &resource : resources_) {
+    CUDA_CALL(cudaStreamSynchronize(resource.stream));
+
+    CUDA_CALL(nvjpegJpegStreamDestroy(resource.jpeg_stream));
+    CUDA_CALL(nvjpegBufferPinnedDestroy(resource.pinned_buffer));
+    CUDA_CALL(nvjpegBufferDeviceDestroy(resource.device_buffer));
+    CUDA_CALL(cudaEventDestroy(resource.decode_event));
+    CUDA_CALL(cudaStreamDestroy(resource.stream));
+
+    CUDA_CALL(nvjpegDecoderDestroy(resource.decoder_data.decoder));
+    CUDA_CALL(nvjpegJpegStateDestroy(resource.decoder_data.state));
+  }
+
+  CUDA_CALL(nvjpegDestroy(nvjpeg_handle_));
+}
+
+void NvJpegDecoderInstance::SetParam(const char *name, const any &value) {
+  if (strcmp(name, "device_memory_padding") == 0) {
+    device_memory_padding_ = any_cast<size_t>(value);
+
+    if (device_memory_padding_ > 0) {
+      for (auto thread_id : tp_->GetThreadIds()) {
+        nvjpeg_memory::AddBuffer<mm::memory_kind::device>(thread_id, device_memory_padding_);
+      }
+    }
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    host_memory_padding_ = any_cast<size_t>(value);
+
+    if (host_memory_padding_ > 0) {
+      for (auto thread_id : tp_->GetThreadIds()) {
+        nvjpeg_memory::AddHostBuffer(thread_id, host_memory_padding_);
+        nvjpeg_memory::AddHostBuffer(thread_id, host_memory_padding_);
+      }
+    }
+  }
+}
+
+any NvJpegDecoderInstance::GetParam(const char *name) const {
+  if (strcmp(name, "device_memory_padding") == 0) {
+    return device_memory_padding_;
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    return host_memory_padding_;
+  } else {
+    DALI_FAIL("Unrecognized param name");
+  }
+}
+
+DecodeResult NvJpegDecoderInstance::DecodeImplTask(int thread_idx,
+                                                   cudaStream_t stream,
+                                                   SampleView<GPUBackend> out,
+                                                   ImageSource *in,
+                                                   DecodeParams opts,
+                                                   const ROI &roi) {
+  DecodingContext ctx = DecodingContext {
+    .resources = resources_[thread_idx],
+  };
+  CUDA_CALL(nvjpegDecodeParamsCreate(nvjpeg_handle_, &ctx.params));
+  CUDA_CALL(nvjpegDecodeParamsSetOutputFormat(ctx.params, GetFormat(opts.format)));
+  CUDA_CALL(nvjpegDecodeParamsSetAllowCMYK(ctx.params, true));
+
+  try {
+    ParseJpeg(*in, opts, ctx);
+    DecodeJpeg(*in, out.mutable_data<uint8_t>(), opts, ctx);
```

Comment:
```suggestion
    DecodeJpegSample(*in, out.mutable_data<uint8_t>(), opts, ctx);
```
I'd rename to something like this to make it clear that this is the sample-by-sample API. `DecoderJpegCuda` would also be OK (as, this is the CUDA based decoder)

Reply:
I'm fine with it. For a batch API I'd expect DecodeJpegs.